gdal/alg/gdalwarpkernel.cpp

```diff
@@ -394,38 +369,40 @@ static void ThreadFuncAdapter(void* pData)
     // Look if we have already a per-thread transformer
     void* pTransformerArg = nullptr;
     const GIntBig nThreadId = CPLGetPID();
-    CPLAcquireMutex(psThreadData->hCondMutex, 1.0);
-    auto oIter = psThreadData->mapThreadToTransformerArg.find(nThreadId);
-    if (oIter != psThreadData->mapThreadToTransformerArg.end())
-    {
-        pTransformerArg = oIter->second;
-    }
-    else if( !psThreadData->bTransformerArgInputAssignedToThread )
+
+
     {
-        // Borrow the original transformer, as it has not already been done
-        psThreadData->bTransformerArgInputAssignedToThread = true;
-        pTransformerArg = psThreadData->pTransformerArgInput;
-        psThreadData->mapThreadToTransformerArg[nThreadId] = pTransformerArg;
+        std::lock_guard<std::mutex> lock(psThreadData->mutex);
+        auto oIter = psThreadData->mapThreadToTransformerArg.find(nThreadId);
+        if (oIter != psThreadData->mapThreadToTransformerArg.end())
+        {
+            pTransformerArg = oIter->second;
+        }
+        else if( !psThreadData->bTransformerArgInputAssignedToThread )
+        {
+            // Borrow the original transformer, as it has not already been done
+            psThreadData->bTransformerArgInputAssignedToThread = true;
+            pTransformerArg = psThreadData->pTransformerArgInput;
+            psThreadData->mapThreadToTransformerArg[nThreadId] = pTransformerArg;
+        }
     }
-    CPLReleaseMutex(psThreadData->hCondMutex);
 
     // If no transformer assigned to current thread, instantiate one
     if( pTransformerArg == nullptr )
     {
+        std::lock_guard<std::mutex> lock(psThreadData->mutex);
```

Comment:
GDALCloneTransformer() might be an expensive operation. It doesn't need to run under the mutex, so you could move the lock just after it

Reply:
That's fine. It wasn't clear from the previous code what was being protected by the mutex. I wasn't sure if it was the `mapThreadToTransformerArg` or the source `pTransformerArgInput` itself or both.

I also assumed the clone would be trivial, so it was better just to protect it as well.